docs/Where-Do-Cookbooks-Come-From.md

```diff
@@ -0,0 +1,73 @@
+## Providing Cookbooks to Chef Server
+
+There are three methods for getting cookbooks onto the server:
+
+### Knife / Chef Server API
+
+ - knife upload COOKBOOK/PATH and knife cookbook upload NAME allow cookbooks to be uploaded.
+ - Dependencies are not automatically uploaded - user must do so manually. It's possible to bulk-upload
+   many cookbooks this way.
+
+### Berkshelf
+
+ -  A berksfile provides a convenient method to manage cookbook dependencies. Operator creates a Berksfile
+ from multiple potential sources to local cache
+ - on berks upload, the pinned versions of all cookbooks (direct and dependencies) are uploaded to Chef Server.
+ - dependencies are resolved by default with gecode, though berks offers an option to resolve
+   with a different resolver. This should be safe, but it is possible that this resolver may resolve
+   deps differently than gecode (used by chef-server).  So when a run_list is posted to the server,
+   dependencies may fail to resolve on the client.
+
+
+### Policyfile
+
+- user creates policy file  that specifies cookbook sources, cookbook version constraints, runlist, override and normal attributes.
+- user runs `chef policy install`
+  - pull all required cookbooks into cache from the specified
+  - resolve deps using `solve` gem, same as the alternate to gecode that berks uses.
```

Comment:
Crazy that policy file doesn't use gecode if that's what chef server suses.

Reply:
It makes sense in context - if you're using policyfiles the server never resolves for you, so there's never any chance at discrepancies.